server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java

```diff
@@ -100,33 +108,29 @@ public DesiredBalanceReconciler(ClusterSettings clusterSettings, ThreadPool thre
             value -> this.undesiredAllocationsLogThreshold = value
         );
 
-        unassignedShards = LongGaugeMetric.create(
-            meterRegistry,
+        meterRegistry.registerLongsGauge(
             "es.allocator.desired_balance.shards.unassigned.current",
             "Current number of unassigned shards",
-            "{shard}"
+            "{shard}",
+            this::getUnassignedShards
         );
-        totalAllocations = LongGaugeMetric.create(
-            meterRegistry,
+        meterRegistry.registerLongsGauge(
             "es.allocator.desired_balance.shards.current",
             "Total number of shards",
-            "{shard}"
+            "{shard}",
+            this::getTotalAllocations
         );
-        undesiredAllocations = LongGaugeMetric.create(
-            meterRegistry,
+        meterRegistry.registerLongsGauge(
             "es.allocator.desired_balance.allocations.undesired.current",
             "Total number of shards allocated on undesired nodes excluding shutting down nodes",
-            "{shard}"
+            "{shard}",
+            this::getUndesiredAllocations
         );
```

Comment:
These return a `Closeable` which (in the APM implementation at least) deregisters the metric when closed, we were ignoring it previously when we used the synchronous API and these metrics live for the life of the node so I assume we don't need to worry about it here.